src/Microsoft.Diagnostics.Monitoring.WebApi/EventMonitoringPassthroughStream.cs

```diff
@@ -0,0 +1,254 @@
+        // of repeatedly searching the payload for the field names in _payloadFilter.
+        private Dictionary<int, string> _payloadFilterIndexCache;
+
+
+        /// <summary>
+        /// A stream that can monitor an event stream which is compatible with <see cref="EventPipeEventSource"/>
+        /// for a specific event while also passing along the event data to a destination stream.
+        /// </summary>
+        /// <param name="providerName">The stopping event provider name.</param>
+        /// <param name="eventName">The stopping event name, which is the concatenation of the task name and opcode name, if set. <see cref="TraceEvent.EventName"/> for more information about the format.</param>
+        /// <param name="payloadFilter">A mapping of the stopping event payload field names to their expected values. A subset of the payload fields may be specified.</param>
+        /// <param name="onEvent">A callback that will be invoked each time the requested event has been observed.</param>
+        /// <param name="onPayloadFilterMismatch">A callback that will be invoked if the field names specified in <paramref name="payloadFilter"/> do not match those in the event's manifest.</param>
+        /// <param name="sourceStream">The source event stream which is compatible with <see cref="EventPipeEventSource"/>.</param>
+        /// <param name="destinationStream">The destination stream to write events.</param>
+        /// <param name="bufferSize">The size of the buffer to use when writing to the <paramref name="destinationStream"/>.</param>
+        /// <param name="callOnEventOnlyOnce">If true, the provided <paramref name="onEvent"/> will only be called for the first matching event.</param>
+        /// <param name="leaveDestinationStreamOpen">If true, the provided <paramref name="destinationStream"/> will not be automatically closed when this class is.</param>
+        public EventMonitoringPassthroughStream(
+            string providerName,
+            string eventName,
+            IDictionary<string, string> payloadFilter,
+            Action<TraceEvent> onEvent,
+            Action<TraceEvent> onPayloadFilterMismatch,
+            Stream sourceStream,
+            Stream destinationStream,
+            int bufferSize,
+            bool callOnEventOnlyOnce,
+            bool leaveDestinationStreamOpen) : base()
+        {
+            _providerName = providerName;
+            _eventName = eventName;
+            _onEvent = onEvent;
+            _onPayloadFilterMismatch = onPayloadFilterMismatch;
+            _sourceStream = sourceStream;
+            _payloadFilter = payloadFilter;
+            _callOnEventOnlyOnce = callOnEventOnlyOnce;
+
+            // Wrap a buffered stream around the destination stream
+            // to avoid slowing down the event processing with the data
+            // passthrough unless there is significant pressure.
+            _destinationStream = new BufferedStream(
```

Comment:
You may want to indicate that the destination stream has to be full duplex. For example, if this was a MemoryStream, the reads and writes would likely conflict with each other.